### DISCUSSION TITLE
interactive lifelib

### DISCUSSION THREAD

**Discussion by declann** [lifelib-dev/lifelib#68] [General]
Hello,

Great work on lifelib!

I've seen the project develop for years but quietly - it's really excellent to see the resources, capabilities and discussions evolve!
I appreciate Python a lot for data science type workloads but I've mainly been throwing myself into Javscript for visual and interactive stuff that is easily shared. I develop [calculang](https://calculang.dev) and you'll see *simple* things along those lines there.

I want to integrate calculang with some Python worldly numbers, so I'm throwing myself also into Python and lifelib now.

All the Python environments will work well with Javascript on *some level*, but I wanted to see how lifelib will run in a Pyodide environment (Python but compiled to run in-browser).

Thankfully this was kindof easy! But I guess there is some instability that I hope can be resolved in the long run, maybe by hosting dependencies.

I used a [Python export](https://modelx.io/blog/2023/07/29/export-feature-intro/) of BasicTerm_S, copied the files into my environment and then regular Python import works. I got an error on the modelx version but I guess it could be resolvable - I'm only really interested in the faster one for interactive outputs anyways!

In the notebook I use [IPyWidgets](https://ipywidgets.readthedocs.io/en/stable/) to make an interaction with an Altair visual of the cashflows. I rushed this a bit (black line is the net cashflow), but it's easy to see a consistent shape in the cashflows across any selection of policies.

Below is a link to the notebook ([self deployed](https://jupyterlite.readthedocs.io/en/stable/quickstart/deploy.html)) if someone wants to try. Let me know if you experience issues, and finally, if someone has built interactive outputs for lifelib I'd love to hear about it!

https://declann.github.io/jupyterlite-things/lab/index.html?path=lifelib%2Fbasicterm-interact.ipynb

This is a simple interaction at a policy-level, but it's clearly showing good performance needed for interactive workloads. That might break on lots of aggregate modelling over lots of data, but I guess it's still an interesting environment to work with. (besides being useful for sharing and for informative things like docs or tutorials - if reliable)

Note: since everything runs in the browser you can edit the Python code in this notebook and re-run cells. Other environments are persistent and will keep your changes across sessions. It's *not* a good idea to rely on persistence here and better to use the Download button if you want to keep something! (this gives a regular `ipynb` file for other sessions/notebook environments)

Cheers,

Declan

**Comment by declann** [comment]
Here's a video, since I'm aware the notebook might not load

[Screencast from 03-04-24 20:17:10.webm](https://github.com/lifelib-dev/lifelib/assets/353214/4785df2e-b6bb-446b-9f97-798299f0e3cf)

**Reply by fumitoh** [reply]
calculang looks interesting. lifelib and modelx are built specific to CPython, so they may not work with Pyodide, because, for one thing, modelx checks `sys.platform` at a few places in its source code. But your interactive notebook doesn't need to use Pyodide, right?
If that's the case, then it should run on Google Colab on-the-fly like notebooks included in lifelib, so
with a few adjustments the notebook could be a good example to be included in lifelib that illustrate how you can use IPyWidgets and Altair with lifelib models.

**Reply by declann** [reply]
Indeed it doesn't need to use Pyodide; it works in other environments locally or else on Binder or Colab. If I get a chance perhaps I'll submit more interactions that you could include in whichever environment you prefer.

Working via Pyodide is also interesting to embed interactions directly into a webpage like a blog (or a report or a dashboard (though Shiny and other things make this already possible)).

If modelx/lifelib on Pyodide isn't a runner, then it's another reason to be thankful for the export functionality !

10k policies on this simple model is a push for *real-time feedback* outside of a fully blown Python environment, but higher-level modelling/model points isn't. Tonnes of things can be shown with modelling only on static values that come from Python models like lifelib, and it's things along those lines that I'll see if I can show together with calculang.

Thanks a lot for the great work you do on modelx and lifelib!

**Comment by declann** [comment]
On my blog I posted a [linked visualization of values from BasicTerm_SE model](https://calcwithdec.dev/posts/linked-visuals/) (excluding NB and reversing scaling by policy_count).

Data is static for this but the visualization is interactive - I mostly wanted to build something using Mosaic which is a really nice framework for this kind of linked visualization and uses DuckDB database to make it scaleable.

It helped me to understand the model through the data also, although there isn't a lot to discover there given pricing criteria very direct relationship to net premium on term and age at entry.

Mosaic has a [Jupyter widget](https://uwdata.github.io/mosaic/jupyter/), so can be integrated in an interactive notebook also.

**Reply by fumitoh** [reply]
Excellent work! 👍

**Comment by declann** [comment]
I took the modelx option pricing example and copied it into a marimo reactive notebook and added some UI elements.

I use reactive javascript a lot (Mike Bostocks observable js/hq) so glad to play with similar reactive systems in python world.

https://marimo.app/l/1cvggw

[Screencast from 12-04-24 22:32:16.webm](https://github.com/lifelib-dev/lifelib/assets/353214/802833bc-ea7c-40c2-8635-7f6671572c3a)

Also started to do the same with shinylive but didn't polish it to the same extent. Shinylive embeds nicely in [Quarto](https://quarto.org) websites though (like my website/blog) - Shiny and Quarto both being Posit projects.

In both shinylive/marimo cases running modelx through pyodide.

Pyodide seems to be pretty robust.

I guess that lifelib will work? in the worst case a serializer to a modelx-python syntax (like above) is *possible*? (i.e. as opposed to requiring a custom parser)

but export to regular Python certainly works (since thats what I used to run in Jupyterlite).

**Reply by fumitoh** [reply]
Cool stuff!

modelx may or may not run on Pyodide, because it uses some CPython-specific techniques.
Further analysis could be possible if error messages are available.

**Comment by declann** [comment]
Marimo has a few limitations but I'm seeing it improve a lot

Below is a link to a notebook which: in WebAssembly
- installs lifelib and dependencies
- exports a basic term model (this will be better when I can just bundle the resulting nomx model - not you have to be a little patient while the above steps complete)
- does some interactive visualization

Very like what I did in Jupyterlite before, but now with the benefit of Marimos reactivity.

[link](https://marimo.app/#code/JYWwDg9gTgLgBCAhlUEBQaD6mDmBTAOzykRjwBNMB3YGACzgF44AiABgDoBODgRi4C05PADcAzCzSIwYJgmSoOAQRkAKAJQY0AAWlgOAYzwAbY2mEAzONlVgATOoBcaOK7j2XbqHhgBXKARaujKGJmaW1piqAEbIAM4ANHDGwERxTp6usVBxANQpaZlw3n4BQXqhpuZ4VjYgEBluCBAcIOSqUCxdAMRwKmCE5MAAHl0smk0l-oEYwfpGVYhxAJ4EBnARNo1uoJCwCMAGUBBgwGBFu9DwKRYmwNFFN3fRht6keKossXGHTyxJLBAyz+EzciCoiFoByOJzOHFScRgiFMqgA5CdCGBlsNjKjQa5Lvt6sJjMM4EsEMMiiBhhw3pRiSY0RwAPRAp4sgBCS0OABViCBMABlFl4jh4YZ7GBo7k-Az8qCCoWYAgQGl4ooWY4gOCyvkC4UqtVkwnwPXyg1CopTAK6nkWxXCpJPFLRJIgQ7HU5gd1U2YVBbhGqRDTOJqm8nGJHAKDkuKRmAXcBXOAEXzgZZx1PnSY+aYJpIEc7+kKB6q1KLmhVK7YE5P7MCIAjkJYYJqQFIGZaYbxxXxR+PMRvNpYcAAipEQADESCAPgBtOCqdxJKuWjgABWOACs8AYYMAIAR52AALocRD4TCkTCEGBQZYaVf26vCzc7vcHo8n89xCVPu05VfIV3wgXd90PY8zw4DtDm7MhFQAtdHRArcwM-SCfw4PtBSWPtvHadRnyA9c0PAr8oPPMARBVHxMAMCwAPUOALGgdw4FSYom3wVReASXg2GY08kgMCBjHTAg4kYedUTAVEElRS88GvGBbwIe9lnk1E-2GLTYK7TAEJALScOvOJ8IoPSaKIVSGNRU98WKPNbWHFtEncMS4J7PA+wHcpSzCcsQ2RGAkn07te37GB0jDMEow4FskUMkhJNYxViDiBLgDiRBomMZSkGGHsICoOINDbOKYA4ABhOhkGlcLvN86L1Bgr1iAPHzVBoch6EYAAWAAONgkjoPBgBwOgYEYOw2BGuADxgfLGFRZCdTQixaFy4AUhgTNokzDdPK7JdtDgCBfGinwAG44AEgBrDzUhasUkCge7MEgZ6NHFNYIGEVQiiaYG3GGFbTLw-wKEcABFeSgZB4HlhW6jaNsixYfhxHsbgbpROMaAVqUlS1I0zGEgRnG4HxwmQpqsToDRYmbzvB9MdauIDGRD5ObGudwYGfcSFxIjKZxuI6oGcGJUcAA5LGqeB44qBR474IFOWFcVtxoGEKAUdRMWQccm0Zh0ANAs2SsX0tQsfXcWsNisZg1rfMiMO-aDIqjejGNa3s6NSYRhh+uco1UYBKBEeJpNRIOJXspJo6gFVEH51EuYliwCaoDUmnICwZOj8S8HsuQiw4Kgxu8VQC5kzO6Gzkqy4AQmYVEtzwD10zieSnaL5FfFL08ACoBD4uvUWLof7Mcyfp+H8v9Cr4gPknhum9z08mHb2WfDgaqlkbnO+-nwfh7HifC6n8-Z6KboC7kOv16PzfW93-fD6zk-T2tZyCCXAXBImgSzzEtsGGwIUkgF0dt8PAcg6a1XqrXCwrUwDtVgMALqPU+oAFZhqjXGpNaa-U5ogKaNkQccA4GtGQB9bIP17xNjiGlQUm0ozEFUOwhCjAWCJXTIYV+Oc4BtzgKiPe8Av7H2buMX6okAZGzcDACAYkDxgFjhvHOmsFIL3JnHZsMt5bCUUa4MGdMAAaaJ47DHJnATm3NGB0yFFzfKtc1SQgINJEas02AOSSItZaLADxznGBTbWyM6YAE00S6Lhkkexy0nEuLXu41I0kBC4LmgkMQc0HKi21jTfWGchElU1oo8hbgCg+TkDQt6H0qkR3UsQDs7xwZkDAAIRAFgEJiiYalaAbCdoIS4UM4gvD+EgEEd-EqO8xESIPiU3OsjCDyI+CYhaKioxnFjvdPAyxtE3xLno6xmtjHazMfFSx+jg62ISXgRx8VnHczcUgNJ85vG5KIgtWggTgl4FCesiJ8VomHKHrc5JDyqpPNceQVJnj5wZKyTk3x6h8mK0KZCjgC80R5UQAYe6eccYm3-tQ+ISQ4HOlSD5fyYCqhW1DEmKU8gUD1CzPUP+pQAGqHqMAmllQgwVm5Q0WKrh7ByHqBwXwwBsIpD1qoRE9VGB8Tsco9RAk5rzURHgNVxLOVLnsLy0B-Kgo2HsI7WQQ47BYvPhy-MthDVoGABWAgadlKYFmSwbAryCDYBYCK8kIQoC+AIOVIAA)
